tests/test_cli_compile.py

```diff
@@ -1534,3 +1534,32 @@ def test_duplicate_reqs_combined(
     assert out.exit_code == 0, out
     assert str(test_package_2) in out.stderr
     assert "test-package-1==0.1" in out.stderr
+
+
+@pytest.mark.network
+def test_triple_equal_pinned_depency_is_used(runner, make_package, make_wheel, tmpdir):
+    """
+    Test that pip-compile properly emits the pinned requirement with ===
+    torchvision 0.8.2 requires torch==1.7.1 which can resolve to versions with
+    patches (e.g. torch 1.7.1+cu110), we want torch===1.7.1 without patches
+    """
+
+    dists_dir = tmpdir / "dists"
+
+    test_package_1 = make_package("test_package_1", version="1.7.1")
+    make_wheel(test_package_1, dists_dir)
+
+    test_package_2 = make_package(
+        "test_package_2", version="0.8.2", install_requires=["test-package-1==1.7.1"]
+    )
+    make_wheel(test_package_2, dists_dir)
+
+    with open("requirements.in", "w") as reqs_in:
+        reqs_in.write("test-package-1===1.7.1\n")
+        reqs_in.write("test-package-2===0.8.2\n")
```

Comment:
Could you delete this line with `test-package-2` pin? Would be nice to test, that the resolver keeps dependency's pin operator?

Reply:
You mean pinning test-package-2 with == 0.8.2 and check that it is kept ?

Reply:
Yeah, left only one pin in `requirements.in`:

```
# requirements.in
test-package-1===1.7.1
```

Reply:
well that defeats the purpose of this particular test :) the issue was that the requirement from test_package_2 which pinned test_package_1 with == was taken instead of the more stringent requirement for test_package_1 with ===, plus test package 1 here does not depend on test_package_2

I can add additional test cases though if you want so that we can verify combinations of pinnings ?

Something like that for example :

```python
    # Case 1 ===
    with open("requirements.in", "w") as reqs_in:
        reqs_in.write("test-package-1===1.7.1\n")

    out = runner.invoke(cli, ["--find-links", str(dists_dir)])

    assert out.exit_code == 0, out
    assert "test-package-1===1.7.1" in out.stderr

    # Case 2 ==
    with open("requirements.in", "w") as reqs_in:
        reqs_in.write("test-package-1==1.7.1\n")

    out = runner.invoke(cli, ["--find-links", str(dists_dir)])

    assert out.exit_code == 0, out
    assert "test-package-1==1.7.1" in out.stderr

    # Case 3 pinned by user with === but pinned by package with ==, prefer ===
    with open("requirements.in", "w") as reqs_in:
        reqs_in.write("test-package-1===1.7.1\n")
        reqs_in.write("test-package-2==0.8.2\n")

    out = runner.invoke(cli, ["--find-links", str(dists_dir)])

    assert out.exit_code == 0, out
    assert "test-package-1===1.7.1" in out.stderr
    assert "test-package-2==0.8.2" in out.stderr

    # Case 4 pinned by user with === but pinned by package with ==, prefer ===
    # Different pin for test_package_2
    with open("requirements.in", "w") as reqs_in:
        reqs_in.write("test-package-1===1.7.1\n")
        reqs_in.write("test-package-2===0.8.2\n")

    out = runner.invoke(cli, ["--find-links", str(dists_dir)])

    assert out.exit_code == 0, out
    assert "test-package-1===1.7.1" in out.stderr
    assert "test-package-2===0.8.2" in out.stderr
```

Reply:
This set of test cases come to my mind:

First round where dependency is `test-package2==0.1`:
```
setup(name='test-package-1', version='0.1', install_requires=['test-package2==0.1'])
```

1. Test direct dependency keeps `===`
```
# requirements.in
test-package-1===0.1
```

2. Test sub-dependency keeps `===`
```
# requirements.in
test-package-1
test-package-2===0.1
```

Second round where dependency is `test-package2===0.1`:

```
setup(name='test-package-1', version='0.1', install_requires=['test-package2===0.1'])
```

1. Test sub-dependency keeps `===`
```
# requirements.in
test-package-1
```

2. Test sub-dependency keeps `==`
```
test-package1==0.1
```

Well, that pretty close to what you've provided, but slightly different on `setup.py` files 👍🏼 Could you also split each test case into a parametrized test?

Reply:
Latest commit tests what you propose except the last case as pip cannot handle package where an install_requires string has a dependency pinned with === apparently, there is a parse error happening in pip._vendor.pyparsing. Is this material for a bug report for pip or should dependencies not be pinned with === in packages ?

Reply:
Ok for some reason the bug is not happening anymore, I'll try to handle the sub-dependency case which does not preserve a package dependency specified with ===

Reply:
> pip does not support === for package dependencies at all...

Oh, interesting. Does it? I've tried it locally though:

```python
# setup.py
from setuptools import setup
setup(name='foo', version='0.1', install_requires=['six===1.10.0'])
```

```console
$ pip install -e .
Obtaining file:///private/var/folders/x4/88j2h7255_qd_td8wsn642br0000gp/T/tmp.95T5QMTv
Collecting six===1.10.0
  Using cached six-1.10.0-py2.py3-none-any.whl (10 kB)
Installing collected packages: six, foo
  Running setup.py develop for foo
Successfully installed foo six-1.10.0
WARNING: You are using pip version 20.1; however, version 21.0.1 is available.
```

Reply:
Yep, "funnily" install -e must use a different route, it worked for a project I'm working on. Try installing a wheel generated from that setup.py, that's where I was out of luck

Reply:
> Yep, "funnily" install -e must use a different route, it worked for a project I'm working on. Try installing a wheel generated from that setup.py, that's where I was out of luck

I see. I've caught this error with wheels:

```
pip._vendor.pkg_resources.RequirementParseError: Parse error at "'(===1.10'": Expected stringEnd
```

> is it worth opening an issue on the pip repo at this point?

Yeah, It'll be good to know what would pip-devs think about it.

Reply:
Issue opened here: https://github.com/pypa/pip/issues/9650

I'll add ids to tests